File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        numLen = len(nums)
        ret = []
        
        if numLen < 3: return ret
        
        nums.sort()
        
        i = 0
        while i < numLen - 2:
            
            j, k = i + 1, numLen - 1
            
            while j < k:
                s = nums[j] + nums[k] + nums[i]
                
                if s == 0:
                    ret.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while k > 0 and nums[k] == nums[k + 1]: k -= 1
                    while j < numLen and nums[j] == nums[j - 1]: j += 1
                elif s > 0:
                    k -= 1
                    while k > 0 and nums[k] == nums[k + 1]: k -= 1
                else:
                    j += 1
                    while j < numLen and nums[j] == nums[j - 1]: j += 1

            i += 1
            while i < numLen and nums[i] == nums[i - 1]: i += 1
                    
        return ret
            
            
        
        
        

File: 15-3sum/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (nums.size() < 3) return {};
        sort(nums.begin(), nums.end());
        vector<vector<int>> ret;
        
        for (int i = 0; i < nums.size() - 2; i++) {
            if (i && nums[i] == nums[i - 1]) continue;
            
            int left = i + 1, right = nums.size() - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    ret.push_back({nums[i], nums[left], nums[right]});
                    left++; 
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                }
                else if (sum > 0) {
                    right--;
                }
                else {
                    left++;
                }
            }
        }
        return ret;
    }
};